test/e2e/tests/e2e_test.go

```diff
@@ -155,3 +161,68 @@ func fetchBlockChain(t *testing.T) []*types.Block {
 
 	return blocks
 }
+
+// fetchABCIRequests go through the logs of a specific node and collect all ABCI requests (each slice represents requests from beggining until the first crash,
+// and then between two crashes) for a specific node.
+func fetchABCIRequests(t *testing.T, nodeName string) ([][]*abci.Request, error) {
+	testnet := loadTestnet(t)
+	logs, err := fetchNodeLogs(testnet, nodeName)
+	if err != nil {
+		return nil, err
+	}
+	reqs := make([][]*abci.Request, 0)
+	i := -1
+	// Parse output line by line.
+	lines := strings.Split(string(logs), "\n")
+	for _, line := range lines {
+		if strings.Contains(line, "Application started") {
+			i += 1
+			reqs = append(reqs, make([]*abci.Request, 0))
+			continue
+		}
+		r, err := app.GetABCIRequestFromString(line)
+		require.NoError(t, err)
+		// Ship the lines that does not contain abci request.
+		if r == nil {
+			continue
+		}
+		reqs[i] = append(reqs[i], r)
+	}
+	return reqs, nil
+}
+
+func fetchNodeLogs(testnet e2e.Testnet, nodeName string) ([]byte, error) {
+	dir := filepath.Join(testnet.Dir, "docker-compose.yml")
+	c1 := exec.Command("docker-compose", "-f", dir, "logs")
+	c2 := exec.Command("grep", nodeName)
```

Comment:
It's probably easier to do the "grep" part internally with `string.Contains`

Reply:
True. I had a reason why I did like this instead but I forgot it. I will change it to `string.Contains` and maybe the reason will appear otherwise I will keep `string.Contains` :D